File: src/main/java/me/zeroeightsix/kami/module/modules/misc/CustomChat.java
package me.zeroeightsix.kami.module.modules.misc;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.network.play.client.CPacketChatMessage;

import static me.zeroeightsix.kami.KamiMod.*;

/**
 * Created by 086 on 8/04/2018.
 * Updated by S-B99 on 26/01/20
 */
@Module.Info(name = "CustomChat", category = Module.Category.MISC, description = "Chat ending", showOnArray = Module.ShowOnArray.OFF)
public class CustomChat extends Module {

    public Setting<Boolean> startupGlobal = register(Settings.b("Enable Automatically", true));
    private Setting<TextMode> textMode = register(Settings.e("Content", TextMode.JAPANESE));
    private Setting<DecoMode> decoMode = register(Settings.e("Punctuation", DecoMode.CLASSIC));
    private Setting<Boolean> commands = register(Settings.b("Commands", false));

    private enum TextMode {
        NAME, ONTOP, WEBSITE, JAPANESE;
    }

    private enum DecoMode {
        SEPARATOR, CLASSIC, NONE
    }

    private String getText(TextMode t) {
        switch (t) {
            case NAME: return KAMI_BLUE;
            case ONTOP: return KAMI_ONTOP;
            case WEBSITE: return KAMI_WEBSITE;
            case JAPANESE: return KAMI_JAPANESE_ONTOP;
            default: return "";
        }
    }

    private String getFull(DecoMode d) {
        switch (d) {
            case NONE: return " " + getText(textMode.getValue());
            case CLASSIC: return  " " + quoteLeft + " " + getText(textMode.getValue()) + " " + quoteRight;
            case SEPARATOR: return " " + separator + " " + getText(textMode.getValue());
            default: return "";
        }
    }

    @EventHandler
    public Listener<PacketEvent.Send> listener = new Listener<>(event -> {
        if (event.getPacket() instanceof CPacketChatMessage) {
            String s = ((CPacketChatMessage) event.getPacket()).getMessage();
            if (!commands.getValue()) {
                if (s.startsWith("/")) return;
                else if (s.startsWith(",")) return;
                else if (s.startsWith(".")) return;
                else if (s.startsWith("-")) return;
                else if (s.startsWith(";")) return;
                else if (s.startsWith("?")) return;
                else if (s.startsWith("*")) return;
                else if (s.startsWith("^")) return;
                else if (s.startsWith("&")) return;
            }
            s += getFull(decoMode.getValue());
            if (s.length() >= 256) s = s.substring(0, 256);
            ((CPacketChatMessage) event.getPacket()).message = s;
        }
    });

}

File: src/main/kotlin/org/kamiblue/client/module/modules/chat/CustomChat.kt
package org.kamiblue.client.module.modules.chat

import net.minecraftforge.fml.common.gameevent.TickEvent
import org.kamiblue.client.manager.managers.MessageManager.newMessageModifier
import org.kamiblue.client.module.Category
import org.kamiblue.client.module.Module
import org.kamiblue.client.util.TickTimer
import org.kamiblue.client.util.TimeUnit
import org.kamiblue.client.util.text.MessageDetection
import org.kamiblue.client.util.text.MessageSendHelper
import org.kamiblue.client.util.threads.safeListener
import kotlin.math.min

internal object CustomChat : Module(
    name = "CustomChat",
    category = Category.CHAT,
    description = "Add a custom ending to your message!",
    showOnArray = false,
    modulePriority = 200
) {
    private val textMode by setting("Message", TextMode.JAPANESE)
    private val decoMode by setting("Separator", DecoMode.NONE)
    private val commands by setting("Commands", false)
    private val spammer by setting("Spammer", false)
    private val customText by setting("Custom Text", "Default")

    private enum class DecoMode {
        SEPARATOR, CLASSIC, NONE
    }

    private enum class TextMode {
        NAME, ON_TOP, WEBSITE, JAPANESE, CUSTOM
    }

    private val timer = TickTimer(TimeUnit.SECONDS)
    private val modifier = newMessageModifier(
        filter = {
            (commands || MessageDetection.Command.ANY detectNot it.packet.message)
                && (spammer || it.source !is Spammer)
        },
        modifier = {
            val message = it.packet.message + getFull()
            message.substring(0, min(256, message.length))
        }
    )

    init {
        onEnable {
            modifier.enable()
        }

        onDisable {
            modifier.disable()
        }
    }

    private fun getText() = when (textMode) {
        TextMode.NAME -> "ᴋᴀᴍɪ ʙʟᴜᴇ"
        TextMode.ON_TOP -> "ᴋᴀᴍɪ ʙʟᴜᴇ ᴏɴ ᴛᴏᴘ"
        TextMode.WEBSITE -> "ｋａｍｉｂｌｕｅ．ｏｒｇ"
        TextMode.JAPANESE -> "上にカミブルー"
        TextMode.CUSTOM -> customText
    }

    private fun getFull() = when (decoMode) {
        DecoMode.NONE -> " " + getText()
        DecoMode.CLASSIC -> " \u00ab " + getText() + " \u00bb"
        DecoMode.SEPARATOR -> " | " + getText()
    }

    init {
        safeListener<TickEvent.ClientTickEvent> {
            if (timer.tick(5L) && textMode == TextMode.CUSTOM && customText.equals("Default", ignoreCase = true)) {
                MessageSendHelper.sendWarningMessage("$chatName Warning: In order to use the custom $name, please change the CustomText setting in ClickGUI")
            }
        }
    }


}
